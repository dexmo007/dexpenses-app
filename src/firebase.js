import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/storage';
import 'firebase/firestore';

const config = {
  apiKey: 'AIzaSyCR0QS9ZNiuBIjv70wl3m1PJrkWrDVxrEU',
  authDomain: 'dexpenses-207219.firebaseapp.com',
  databaseURL: 'https://dexpenses-207219.firebaseio.com',
  projectId: 'dexpenses-207219',
  storageBucket: 'dexpenses-207219.appspot.com',
  messagingSenderId: '45657867106',
};
firebase.initializeApp(config);
