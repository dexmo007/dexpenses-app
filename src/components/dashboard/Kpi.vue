<template>
  <div class="kpi-container">
    <h2 v-if="title">{{title}}</h2>
    <span v-if="value || value === 0">{{value}} â‚¬</span>
    <span
      class="red--text text--lighten-1"
      v-else-if="error"
    >Error</span>
    <div
      class="loader"
      v-else
    >
      <v-progress-circular indeterminate />
    </div>
  </div>
</template>
<script>
import firebase from 'firebase/app';

export default {
  props: {
    title: String,
    func: {
      type: String,
      required: true,
    },
    data: Object,
  },
  data() {
    return {
      value: null,
      error: false,
    };
  },
  async mounted() {
    try {
      const { data } = await firebase.functions().httpsCallable(this.func)(
        this.data || {}
      );
      this.value = data.value || 0;
    } catch {
      this.error = true;
    }
  },
};
</script>
<style scoped>
.kpi-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  align-items: center;
  justify-content: center;
}
.kpi-container span {
  font-weight: bold;
  font-size: 48px; /* TODO: responsive (vw?) */
}
.loader {
  height: 72px;
  display: flex;
  align-items: center;
}
</style>
