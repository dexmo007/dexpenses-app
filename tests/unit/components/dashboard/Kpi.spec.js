import { shallowMount } from '@vue/test-utils';
import firebase from 'firebase/app';
import Kpi from '@/components/dashboard/Kpi.vue';

jest.mock('firebase/app', () => ({
  functions: jest.fn().mockReturnValue({
    httpsCallable: jest.fn().mockReturnValue(jest.fn().mockResolvedValue({ data: { value: 1 } })),
  }),
}));

describe('dashboard/Kpi', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  function shallowMountComponent() {
    return shallowMount(Kpi, {
      propsData: {
        title: 'title',
        func: 'mockFunc',
      },
      stubs: {
        'v-progress-circular': true,
      },
    });
  }
  it('should render value correctly', async () => {
    const wrapper = shallowMountComponent();
    expect(wrapper.element).toMatchSnapshot(); // shows loading
    await new Promise(r => setTimeout(r));
    expect(wrapper.element).toMatchSnapshot(); // shows value
  });

  it('should render absent value correctly', async () => {
    firebase
      .functions()
      .httpsCallable('mockFunc')
      .mockResolvedValue({ data: { value: null } });
    const wrapper = shallowMountComponent();
    expect(wrapper.element).toMatchSnapshot(); // shows loading
    await new Promise(r => setTimeout(r));
    expect(wrapper.element).toMatchSnapshot(); // shows 0 as value
  });

  it('should render error correctly', async () => {
    firebase
      .functions()
      .httpsCallable('mockFunc')
      .mockRejectedValue();
    const wrapper = shallowMountComponent();
    expect(wrapper.element).toMatchSnapshot(); // shows loading
    await new Promise(r => setTimeout(r));
    expect(wrapper.element).toMatchSnapshot(); // shows error
  });
});
