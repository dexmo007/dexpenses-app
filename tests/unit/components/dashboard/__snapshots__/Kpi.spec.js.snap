// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard/Kpi should render absent value correctly 1`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <div
    class="loader"
  >
    <v-progress-circular-stub
      indeterminate=""
    />
  </div>
</div>
`;

exports[`dashboard/Kpi should render absent value correctly 2`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <span>
    0 €
  </span>
</div>
`;

exports[`dashboard/Kpi should render error correctly 1`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <div
    class="loader"
  >
    <v-progress-circular-stub
      indeterminate=""
    />
  </div>
</div>
`;

exports[`dashboard/Kpi should render error correctly 2`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <span
    class="red--text text--lighten-1"
  >
    Error
  </span>
</div>
`;

exports[`dashboard/Kpi should render value correctly 1`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <div
    class="loader"
  >
    <v-progress-circular-stub
      indeterminate=""
    />
  </div>
</div>
`;

exports[`dashboard/Kpi should render value correctly 2`] = `
<div
  class="kpi-container"
>
  <h2>
    title
  </h2>
   
  <span>
    1 €
  </span>
</div>
`;
